File: sdk-go/flowsdk/v1beta2/interaction.go
package v1beta2

import (
	flowv1beta2 "github.com/datakit-dev/dtkt-sdk/sdk-go/proto/dtkt/flow/v1beta2"
	protoformv1beta1 "github.com/datakit-dev/dtkt-sdk/sdk-go/proto/dtkt/protoform/v1beta1"
	form "github.com/datakit-dev/dtkt-sdk/sdk-go/protoformsdk/v1beta1"
	"google.golang.org/protobuf/proto"
)

// GetInteractionBinding returns the protoform FieldElement and binding proto
// for an interaction input. The binding's "value" field is used to render a
// form widget and collect user input via protoformsdk.
//
// Returns (nil, nil, false) if the input's element type is unrecognised.
func GetInteractionBinding(input *flowv1beta2.Interaction_Input) (*protoformv1beta1.FieldElement, proto.Message, bool) {
	var binding proto.Message
	switch input.GetElement().(type) {
	case *flowv1beta2.Interaction_Input_Confirm:
		binding = &flowv1beta2.Interaction_ConfirmBinding{}
	case *flowv1beta2.Interaction_Input_Input:
		binding = &flowv1beta2.Interaction_InputBinding{}
	case *flowv1beta2.Interaction_Input_File:
		binding = &flowv1beta2.Interaction_FileBinding{}
	case *flowv1beta2.Interaction_Input_Select:
		binding = &flowv1beta2.Interaction_SelectBinding{}
	case *flowv1beta2.Interaction_Input_MultiSelect:
		binding = &flowv1beta2.Interaction_MultiSelectBinding{}
	}

	if binding == nil {
		return nil, nil, false
	}

	desc := binding.ProtoReflect().Descriptor().Fields().ByName("value")
	elem, ok := form.GetFieldElement(desc)
	if !ok {
		return nil, nil, false
	}

	title := input.GetTitle()
	elem.Title = &title
	if input.HasDescription() {
		d := input.GetDescription()
		elem.Description = &d
	}

	// Merge element-specific options from the spec into the form element.
	switch elemType := input.GetElement().(type) {
	case *flowv1beta2.Interaction_Input_Confirm:
		if elemType.Confirm != nil && elem.GetConfirm() != nil {
			proto.Merge(elem.GetConfirm(), elemType.Confirm)
		} else if elemType.Confirm != nil {
			elem.Type = &protoformv1beta1.FieldElement_Confirm{Confirm: elemType.Confirm}
		}
	case *flowv1beta2.Interaction_Input_Input:
		if elemType.Input != nil {
			elem.Type = &protoformv1beta1.FieldElement_Input{Input: elemType.Input}
		}
	case *flowv1beta2.Interaction_Input_Select:
		if elemType.Select != nil {
			elem.Type = &protoformv1beta1.FieldElement_Select{Select: elemType.Select}
		}
	case *flowv1beta2.Interaction_Input_MultiSelect:
		if elemType.MultiSelect != nil {
			elem.Type = &protoformv1beta1.FieldElement_MultiSelect{MultiSelect: elemType.MultiSelect}
		}
	}

	return elem, binding, true
}

File: sdk-ts/src/flowsdk/v1beta2/interaction.ts
import { create, type DescField, type DescMessage, getOption, hasOption, type Message } from "@bufbuild/protobuf";

import { Interaction_ConfirmBindingSchema, Interaction_FileBindingSchema, type Interaction_Input, Interaction_InputBindingSchema, Interaction_MultiSelectBindingSchema, Interaction_SelectBindingSchema } from "../../proto/dtkt/flow/v1beta2/spec_pb";
import { field as fieldExt, type FieldElement, FieldElementSchema } from "../../proto/dtkt/protoform/v1beta1/protoform_pb";

export type InteractionBindingResult = {
  /**
   * The protoform FieldElement driving the widget. Lifted from the
   * binding's `value` field's annotation, then overlayed with the
   * input's title / description and any element-specific options
   * the spec carried (ConfirmElement labels, InputElement multiline
   * flag, Select/MultiSelect choices, etc.).
   */
  element: FieldElement;
  /**
   * The empty binding proto (ConfirmBinding / InputBinding / FileBinding
   * / SelectBinding / MultiSelectBinding) whose `value` field is what
   * the user fills in.
   */
  binding: Message;
  /**
   * Descriptor for the binding -- consumers `reflect(desc, binding)`
   * for protoform rendering or `toJson(desc, binding)` when packing
   * the response struct.
   */
  desc: DescMessage;
};

/**
 * Returns the protoform FieldElement and binding proto for an
 * interaction input. Mirrors `GetInteractionBinding` in
 * `sdk-go/flowsdk/v1beta2/interaction.go`.
 *
 * The binding's `value` field carries a `(dtkt.protoform.v1beta1.field)`
 * extension declaring the widget kind (confirm / input / file / select /
 * multi_select). We lift that annotation, overlay the input's title /
 * description, and merge any element-specific options the spec set
 * (button labels for Confirm, multiline for Input, choices for Select /
 * MultiSelect). The resulting FieldElement drives the widget; the
 * empty binding is the message the user populates and we send back.
 *
 * Returns `undefined` if the element oneof is unset (which can't happen
 * for a well-formed event but defends against forward-compat changes).
 */
export function getInteractionBinding(input: Interaction_Input): InteractionBindingResult | undefined {
  let binding: Message;
  let desc: DescMessage;
  switch (input.element.case) {
    case "confirm":
      binding = create(Interaction_ConfirmBindingSchema);
      desc = Interaction_ConfirmBindingSchema;
      break;
    case "input":
      binding = create(Interaction_InputBindingSchema);
      desc = Interaction_InputBindingSchema;
      break;
    case "file":
      binding = create(Interaction_FileBindingSchema);
      desc = Interaction_FileBindingSchema;
      break;
    case "select":
      binding = create(Interaction_SelectBindingSchema);
      desc = Interaction_SelectBindingSchema;
      break;
    case "multiSelect":
      binding = create(Interaction_MultiSelectBindingSchema);
      desc = Interaction_MultiSelectBindingSchema;
      break;
    default:
      return undefined;
  }

  const valueField: DescField | undefined = desc.field.value;
  let element: FieldElement;
  if (valueField && hasOption(valueField, fieldExt)) {
    element = getOption(valueField, fieldExt);
  } else {
    element = create(FieldElementSchema);
  }

  element.title = input.title;
  if (input.description !== undefined) {
    element.description = input.description;
  }

  // Merge element-specific options from the spec into the FieldElement
  // so the widget receives the Confirm/Input/Select/MultiSelect
  // overrides authored on the interaction.
  switch (input.element.case) {
    case "confirm":
      if (element.type.case === "confirm") {
        element.type = { case: "confirm", value: { ...element.type.value, ...input.element.value } };
      } else {
        element.type = { case: "confirm", value: input.element.value };
      }
      break;
    case "input":
      element.type = { case: "input", value: input.element.value };
      break;
    case "select":
      element.type = { case: "select", value: input.element.value };
      break;
    case "multiSelect":
      element.type = { case: "multiSelect", value: input.element.value };
      break;
    case "file":
      element.type = { case: "file", value: input.element.value };
      break;
  }

  return { element, binding, desc };
}
